src/core/grid_based_algorithms/lb_interface.cpp

```diff
@@ -1273,13 +1265,64 @@ const Vector6d lb_lbnode_get_pi_neq(const Vector3i &ind) {
                              lblattice.halo_grid);
 
     mpi_recv_fluid(node, index, &rho, j, pi.data());
-    // unit conversion
-    p_pi = pi / lbpar.tau / lbpar.tau / lbpar.agrid;
+    p_pi = pi;
 #endif // LB
   }
   return p_pi;
 }
 
+/** calculates the average stress of all nodes by iterating
+ * over all nodes and deviding by the number_of_nodes.
+ */
+const Vector6d lb_lbfluid_get_stress() {
+  Vector6d p{0, 0, 0, 0, 0, 0};
+
+  if (lattice_switch == ActiveLB::GPU) {
+#ifdef LB_GPU
+    // Copy observable data from gpu
+    std::vector<LB_rho_v_pi_gpu> host_values(lbpar_gpu.number_of_nodes);
+    lb_get_values_GPU(host_values.data());
+    std::for_each(host_values.begin(), host_values.end(),
+                  [&p](LB_rho_v_pi_gpu &v) {
+                    for (int i = 0; i < 6; i++)
+                      p[i] += v.pi[i];
+                  });
+
+    // Normalize
+    p *= (1. / lbpar_gpu.number_of_nodes);
+
+    // Add equilibrium stress to the diagonal (in LB units)
+    double p0 = lb_lbfluid_get_density() * lbmodel.c_sound_sq;
+
+    p[0] += p0;
+    p[2] += p0;
+    p[5] += p0;
+
+#endif
+  } else
+#ifdef LB
+      if (lattice_switch == ActiveLB::CPU) {
+    int number_of_nodes = lblattice.global_grid[0] * lblattice.global_grid[1] *
+                          lblattice.global_grid[2];
+
+    for (int i = 0; i < lblattice.global_grid[0]; i++) {
+      for (int j = 0; j < lblattice.global_grid[1]; j++) {
+        for (int k = 0; k < lblattice.global_grid[2]; k++) {
+          const Vector3i node{{i, j, k}};
+          p += lb_lbnode_get_pi(node);
```

Comment:
this is insanely slow because you have "number of nodes" times the mpi communication. We should think about a generic solution to access all LB nodes

Reply:
See timings in #2054. Walberla has its own efficient implementations, afaik.